packages/kernel/src/client.ts

```diff
@@ -278,12 +289,45 @@ export class LiteKernelClient implements Kernel.IKernelAPIClient {
     return this._kernels.get(id);
   }
 
+  /**
+   * Handle stdin request received from Service Worker.
+   */
+  async handleStdin(
+    inputRequest: KernelMessage.IInputRequestMsg,
+  ): Promise<KernelMessage.IInputReplyMsg> {
+    this._stdinPromise = new PromiseDelegate<KernelMessage.IInputReplyMsg>();
+
+    const clientId = inputRequest.parent_header.session;
+    const kernelId = this._getClientKernel(clientId);
+    if (kernelId !== undefined) {
+      const sendMessage = this._kernelSends.get(kernelId);
+      if (sendMessage !== undefined) {
+        sendMessage(inputRequest);
+      }
+    }
+
+    // Promise is resolved by input reply message.
+    return this._stdinPromise.promise;
+  }
+
+  /**
+   * Get a kernel id corresponding to a client id.
+   */
+  private _getClientKernel(clientId: string): string | undefined {
+    // Walk the _kernelClients to find a match.
+    return this._kernelClients
+      .keys()
+      .find((kernelId) => this._kernelClients.get(kernelId)!.has(clientId));
+  }
+
   private _kernels = new ObservableMap<IKernel>();
   private _clients = new ObservableMap<WebSocketClient>();
   private _kernelClients = new ObservableMap<Set<string>>();
   private _kernelspecs: IKernelSpecs;
   private _serverSettings: ServerConnection.ISettings;
   private _changed = new Signal<this, IObservableMap.IChangedArgs<IKernel>>(this);
+  private _stdinPromise?: PromiseDelegate<KernelMessage.IInputReplyMsg>;
+  private _kernelSends = new ObservableMap<(msg: KernelMessage.IMessage) => void>();
```

Comment:
Here is the mapping for kernel `sendMessage` functions, could be combined with `_kernels` instead?